subprojects/workers/src/main/java/org/gradle/process/internal/daemon/WorkerDaemonExpiration.java

```diff
@@ -0,0 +1,117 @@
+        try {
+            return memoryInfo.getTotalPhysicalMemory();
+        } catch (UnsupportedOperationException ex) {
+            return -1;
+        }
+    }
+
+    @Override
+    public long attemptToRelease(long memoryAmountBytes) throws IllegalArgumentException {
+        if (memoryAmountBytes < 0) {
+            throw new IllegalArgumentException("Negative memory amount");
+        }
+        LOGGER.debug("Will attempt to release {} of memory", memoryAmountBytes / 1024 / 1024);
+        SimpleMemoryExpirationSelector selector = new SimpleMemoryExpirationSelector(memoryAmountBytes);
+        clientsManager.selectIdleClientsToStop(selector);
+        return selector.getReleasedBytes();
+    }
+
+    /**
+     * Simple implementation of memory based expiration.
+     *
+     * Use the maximum heap size of each daemon, not their actual memory usage.
+     * Expire as much daemons as needed to free the requested memory under the threshold.
+     */
+    private class SimpleMemoryExpirationSelector implements Transformer<List<WorkerDaemonClient>, List<WorkerDaemonClient>> {
+
+        private final long memoryBytesToRelease;
+        private long releasedBytes;
+
+        public SimpleMemoryExpirationSelector(long memoryBytesToRelease) {
+            this.memoryBytesToRelease = memoryBytesToRelease;
+        }
+
+        public long getReleasedBytes() {
+            return releasedBytes;
+        }
+
+        @Override
+        public List<WorkerDaemonClient> transform(List<WorkerDaemonClient> idleClients) {
+            List<WorkerDaemonClient> toExpire = new ArrayList<WorkerDaemonClient>();
+            for (WorkerDaemonClient idleClient : idleClients) {
+                toExpire.add(idleClient);
+                long freed = getMemoryUsage(idleClient);
+                releasedBytes += freed;
+                if (releasedBytes >= memoryBytesToRelease) {
+                    break;
+                }
+            }
+            if (LOGGER.isDebugEnabled() && !toExpire.isEmpty()) {
+                // TODO Only log expired workers count, log their "identity" once they are nameable/describable
+                LOGGER.debug("Worker Daemon(s) expired to free some system memory {}", toExpire.size());
+            }
+            return toExpire;
+        }
+
+        private long getMemoryUsage(WorkerDaemonClient idleClient) {
+            // TODO Use actual memory usage received asynchronously from the worker daemon process
+            // For now, use max heap size
+            String forkOptionsMaxHeapSize = idleClient.getForkOptions().getMaxHeapSize();
+            long parsed = MemoryAmount.parseNotation(forkOptionsMaxHeapSize);
+            if (parsed != -1) {
+                // From fork options
+                return parsed;
+            }
+            if (osTotalMemory != -1) {
+                // Calculated based on OS total memory
+                return new MaximumHeapHelper().getDefaultMaximumHeapSize(osTotalMemory);
```

Comment:
The more I think about this, the less I think it makes sense to implement this approximation.  Some of the assumptions (like isIbmJvm or isServerJvm) are based on the current process rather than the worker process.  These really only work by coincidence and the estimate could be wildly off if they are wrong.  It feels like this is not releasable with this approximation and we would only release if we also did the get-memory-from-the-running-process-story.  If that's the case, then I think we should either just combine these stories OR do the get-memory-from-the-running-process-story first.  Doing the other story first makes a lot of sense to me.